File: src/services/processor/intentional_bugs.py
"""
Intentional security bugs for AI review E2E testing.
This file contains vulnerabilities that should be detected by AI code review.
"""
import os
import pickle
import subprocess
import sqlite3


def execute_user_command(user_input: str) -> str:
    """Command injection vulnerability - user input passed directly to shell."""
    result = subprocess.run(user_input, shell=True, capture_output=True, text=True)
    return result.stdout


def unsafe_sql_query(user_id: str) -> list:
    """SQL injection vulnerability - string concatenation in query."""
    conn = sqlite3.connect(":memory:")
    cursor = conn.cursor()
    query = f"SELECT * FROM users WHERE id = '{user_id}'"
    cursor.execute(query)
    return cursor.fetchall()


def deserialize_untrusted(data: bytes) -> object:
    """Insecure deserialization - pickle with untrusted data."""
    return pickle.loads(data)


def hardcoded_credentials():
    """Hardcoded secrets in source code."""
    api_key = "sk-proj-1234567890abcdefghijklmnop"
    db_password = "SuperSecret123!"
    return {"key": api_key, "password": db_password}


def path_traversal(filename: str) -> str:
    """Path traversal vulnerability - no sanitization of user input."""
    base_path = "/var/data/"
    return open(base_path + filename).read()

File: src/services/edge/src/intentional_bugs.rs
//! Intentional security bugs for AI review E2E testing.
//! This file contains vulnerabilities that should be detected by AI code review.

use std::process::Command;
use std::fs;

/// Command injection - user input passed to shell
pub fn execute_command(user_input: &str) -> String {
    let output = Command::new("sh")
        .arg("-c")
        .arg(user_input)  // Direct user input to shell!
        .output()
        .expect("Failed to execute");
    String::from_utf8_lossy(&output.stdout).to_string()
}

/// SQL injection via format string
pub fn unsafe_query(user_id: &str) -> String {
    format!("SELECT * FROM users WHERE id = '{}'", user_id)
}

/// Path traversal - no input validation
pub fn read_file(filename: &str) -> String {
    let path = format!("/var/data/{}", filename);
    fs::read_to_string(path).unwrap_or_default()
}

/// Hardcoded credentials
pub fn get_credentials() -> (&'static str, &'static str) {
    let api_key = "AKIAIOSFODNN7EXAMPLE";
    let secret = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY";
    (api_key, secret)
}

/// Unsafe unwrap that can panic
pub fn parse_config(input: &str) -> i32 {
    input.parse().unwrap()  // Will panic on invalid input
}

/// Use after free potential (unsafe block)
pub unsafe fn dangerous_pointer_math(ptr: *mut i32, offset: isize) -> i32 {
    *ptr.offset(offset)
}
